src/Features/Core/Portable/ExtractMethod/SelectionValidator.cs

```diff
@@ -33,106 +29,98 @@ public abstract partial class SelectionValidator(
 
         public bool ContainsValidSelection => !OriginalSpan.IsEmpty;
 
-        protected abstract SelectionInfo GetInitialSelectionInfo(CancellationToken cancellationToken);
-        protected abstract Task<TSelectionResult> CreateSelectionResultAsync(SelectionInfo selectionInfo, CancellationToken cancellationToken);
+        protected abstract TextSpan GetAdjustedSpan(TextSpan textSpan);
 
-        public abstract ImmutableArray<TExecutableStatementSyntax> GetOuterReturnStatements(SyntaxNode commonRoot, ImmutableArray<SyntaxNode> jumpsOutOfRegion);
-        public abstract bool IsFinalSpanSemanticallyValidSpan(TextSpan textSpan, ImmutableArray<TExecutableStatementSyntax> returnStatements, CancellationToken cancellationToken);
-        public abstract bool ContainsNonReturnExitPointsStatements(ImmutableArray<SyntaxNode> jumpsOutOfRegion);
+        protected abstract SelectionInfo GetInitialSelectionInfo();
 
-        protected abstract bool AreStatementsInSameContainer(TStatementSyntax statement1, TStatementSyntax statement2);
+        protected abstract SelectionInfo UpdateSelectionInfo(
+            SelectionInfo selectionInfo, TStatementSyntax? firstStatement, TStatementSyntax? lastStatement, CancellationToken cancellationToken);
+        protected abstract Task<SelectionResult> CreateSelectionResultAsync(SelectionInfo selectionInfo, CancellationToken cancellationToken);
 
-        public async Task<(TSelectionResult, OperationStatus)> GetValidSelectionAsync(CancellationToken cancellationToken)
+        public async Task<(SelectionResult?, OperationStatus)> GetValidSelectionAsync(CancellationToken cancellationToken)
         {
             if (!this.ContainsValidSelection)
                 return (null, OperationStatus.FailedWithUnknownReason);
 
-            var selectionInfo = GetInitialSelectionInfo(cancellationToken);
+            var selectionInfo = this.GetInitialSelectionInfo();
             if (selectionInfo.Status.Failed)
                 return (null, selectionInfo.Status);
 
+            TStatementSyntax? firstStatement = null;
```

Comment:
this will get more cleaned up in followup.  but the idea here is again just pulling logic up from C#/VB to common location.